adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/MqttRequestTenantAndAuthIdExtractor.java

```diff
@@ -0,0 +1,105 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.mqtt;
+
+import java.net.HttpURLConnection;
+
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.client.TenantClientFactory;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.service.BasicTenantAndAuthIdExtractor;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.eclipse.hono.util.TenantObjectWithAuthId;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.mqtt.MqttAuth;
+import io.vertx.mqtt.MqttEndpoint;
+
+/**
+ * Extracts the tenant and authentication identifier out of an MQTT request.
+ */
+public class MqttRequestTenantAndAuthIdExtractor extends BasicTenantAndAuthIdExtractor {
+
+    /**
+     * Creates a new MqttRequestTenantAndAuthIdExtractor.
+     *
+     * @param config The configuration.
+     * @param tenantClientFactory The factory to use for creating a Tenant service client.
+     * @throws NullPointerException if either of the parameters is {@code null}.
+     */
+    public MqttRequestTenantAndAuthIdExtractor(final ProtocolAdapterProperties config,
+            final TenantClientFactory tenantClientFactory) {
+        super(config, tenantClientFactory);
+    }
+
+    /**
+     * Extracts the tenant and authentication identifier out of either the client certificate or the username/password
+     * credentials provided in the MQTT CONNECT packet.
+     * <p>
+     * This method will return a failed future in case of an unauthenticated MQTT connection.
+     *
+     * @param endpoint The MQTT endpoint to extract the data out of.
+     * @param spanContext The OpenTracing context to use for tracking the operation (may be {@code null}).
+     * @return A future containing the tenant and authentication identifier. The future will be failed if there was an
+     *         error getting the tenant and authentication identifier via the tenantAndAuthIdExtractor. In particular,
+     *         the future will be failed with a {@link ClientErrorException} if there was no corresponding data given in
+     *         the connection attempt to determine the tenant.
+     * @throws NullPointerException if endpoint is {@code null}.
+     */
+    public Future<TenantObjectWithAuthId> getFromMqttEndpoint(final MqttEndpoint endpoint, final SpanContext spanContext) {
+
+        return getFromClientCertificate(endpoint.sslSession(), null, spanContext)
+                .recover(thr -> getFromMqttAuth(endpoint.auth(), spanContext));
+    }
+
+    private Future<TenantObjectWithAuthId> getFromMqttAuth(final MqttAuth auth, final SpanContext spanContext) {
```

Comment:
didn't we want to get rid of the dreaded `TenantObjectWithAuthId` class?
How about passing in the context to set the resulting objects on instead?

Reply:
> How about passing in the context to set the resulting objects on instead?

Then the `MqttContext#tenantObject` and `MqttContext#authId` fields couldn't be final any more. And in my view, having such a method in the extractor class that 'enhances' an `MqttContext` like that wouldn't be very straightforward.
Then I'd rather have the `TenantObjectWithAuthId` class and see it only be used in the *Extractor classes.

As an alternative, we could introduce a `Tuple` class in analogy to the Hono `TriTuple` class and use that instead of `TenantObjectWithAuthId`.

Reply:
I've now pushed a change using `Pair<TenantObject, String>` instead of `TenantObjectWithAuthId`. Once all the adapters have been changed like this, the `TenantObjectWithAuthId` class can be removed then.